File: src/com/company/other/binary_search/BinarySearch.java
package com.company.other.binary_search;

public class BinarySearch {
    public static void main(String[] args) {
        int[] arr = new int[]{-1, 0, 3, 5, 9, 12};
//        System.out.println(search(arr, 2));

//        System.out.println(6 % 4);
        System.out.println(7 / 4);
        System.out.println(7 / 3);
    }

    public static int search(int[] nums, int target) {
        return search(nums, target, 0, nums.length - 1);
    }

    private static int search(int[] nums, int target, int left, int right) {
        if (left > right) return -1;
        int mid = (left + right) / 2;
        int value = nums[mid];
        if (value == target) return mid;

        if (value > target) {
            return search(nums, target, left, mid - 1);
        } else {
            return search(nums, target, mid + 1, right);
        }
    }
}

File: Al_p/algorithms/binary_search.py
def binary_search_rec(sorted_arr, a):
    def search(l, r):
        if l <= r:
            m = (l + r) // 2
            if sorted_arr[m] == a:
                return True
            elif sorted_arr[m] < a:
                return search(m + 1, r)
            else:
                return search(l, m - 1)
        return False

    return search(0, len(sorted_arr) - 1)


def binary_search(arr, a):
    l, r = 0, len(arr) - 1

    while l <= r:
        m = (l + r) // 2
        if arr[m] == a:
            return True
        elif arr[m] < a:
            l = m + 1
        else:
            r = m - 1

    return False


arr = [1, 3, 4, 6, 9]

print(binary_search(arr, 1))
print(binary_search(arr, 2))
print(binary_search(arr, 4))
print(binary_search(arr, 5))
print(binary_search(arr, 9))
